GsonComponentSerializer cannot serialize LeatherArmor hover event with armor trim

### Expected behavior

When we run this code:
```java
        ItemStack item = new ItemStack(Material.LEATHER_CHESTPLATE);
        item.editMeta(ArmorMeta.class, m -> m.setTrim(new ArmorTrim(TrimMaterial.QUARTZ, TrimPattern.SILENCE)));
        Component component = Component.empty().hoverEvent(item.asHoverEvent());
        String string = GsonComponentSerializer.gson().serialize(component);
```
One would expect that the `string` variable now holds a nice representation of the chat component.

### Observed/Actual behavior

Instead it throws and we get this lengthy error message
```
[14:39:19 WARN]: Unexpected exception while parsing console command "test"                                                                                                                                                         [80/1915]
org.bukkit.command.CommandException: Unhandled exception executing command 'test' in plugin Test v0.1
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:164) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]                                                                                                                  
        at org.bukkit.craftbukkit.CraftServer.dispatchCommand(CraftServer.java:964) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                    
        at org.bukkit.craftbukkit.CraftServer.dispatchServerCommand(CraftServer.java:949) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                             
        at net.minecraft.server.dedicated.DedicatedServer.handleConsoleInputs(DedicatedServer.java:521) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                                                              
        at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:468) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                                                                     
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1571) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                                                                                
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1231) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                                                                                 
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:323) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]                                                                                                              
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]                                                    
Caused by: java.lang.IllegalStateException: Failed to convert data component value of type class io.papermc.paper.adventure.PaperAdventure$DataComponentValueImpl to type interface net.kyori.adventure.text.serializer.gson.GsonDataCompone
ntValue due to an error in a converter provided by adventure:platform/paper!
        at net.kyori.adventure.text.event.DataComponentValueConverterRegistry.convert(DataComponentValueConverterRegistry.java:96) ~[adventure-api-4.17.0.jar:?]
        at net.kyori.adventure.text.event.HoverEvent$ShowItem.dataComponentsAs(HoverEvent.java:623) ~[adventure-api-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ShowItemSerializer.write(ShowItemSerializer.java:139) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ShowItemSerializer.write(ShowItemSerializer.java:49) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.StyleSerializer.write(StyleSerializer.java:296) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.StyleSerializer.write(StyleSerializer.java:62) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJsonTree(Gson.java:712) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.ComponentSerializerImpl.write(ComponentSerializerImpl.java:252) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ComponentSerializerImpl.write(ComponentSerializerImpl.java:77) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:812) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:759) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:736) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.GsonComponentSerializerImpl.serialize(GsonComponentSerializerImpl.java:104) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.GsonComponentSerializerImpl.serialize(GsonComponentSerializerImpl.java:41) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at Test.jar/com.cavetale.test.TestPlugin.onCommand(TestPlugin.java:71) ~[Test.jar:?]
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        ... 9 more
Caused by: java.lang.IllegalStateException: Can't access registry ResourceKey[minecraft:root / minecraft:item]; Can't access registry ResourceKey[minecraft:root / minecraft:item]
        at com.mojang.serialization.DataResult$Error.getOrThrow(DataResult.java:275) ~[datafixerupper-7.0.14.jar:?]
        at com.mojang.serialization.DataResult.getOrThrow(DataResult.java:79) ~[datafixerupper-7.0.14.jar:?]
        at io.papermc.paper.adventure.providers.DataComponentValueConverterProviderImpl.lambda$conversions$0(DataComponentValueConverterProviderImpl.java:40) ~[paper-1.20.6.jar:1.20.6-130-5d8e53d]
        at net.kyori.adventure.text.event.DataComponentValueConversionImpl.convert(DataComponentValueConversionImpl.java:59) ~[adventure-api-4.17.0.jar:?]
        at net.kyori.adventure.text.event.DataComponentValueConverterRegistry.convert(DataComponentValueConverterRegistry.java:94) ~[adventure-api-4.17.0.jar:?]
        at net.kyori.adventure.text.event.HoverEvent$ShowItem.dataComponentsAs(HoverEvent.java:623) ~[adventure-api-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ShowItemSerializer.write(ShowItemSerializer.java:139) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ShowItemSerializer.write(ShowItemSerializer.java:49) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.StyleSerializer.write(StyleSerializer.java:296) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.StyleSerializer.write(StyleSerializer.java:62) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJsonTree(Gson.java:712) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.ComponentSerializerImpl.write(ComponentSerializerImpl.java:252) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.ComponentSerializerImpl.write(ComponentSerializerImpl.java:77) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:196) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:842) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:812) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:759) ~[gson-2.10.1.jar:?]
        at com.google.gson.Gson.toJson(Gson.java:736) ~[gson-2.10.1.jar:?]
        at net.kyori.adventure.text.serializer.gson.GsonComponentSerializerImpl.serialize(GsonComponentSerializerImpl.java:104) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at net.kyori.adventure.text.serializer.gson.GsonComponentSerializerImpl.serialize(GsonComponentSerializerImpl.java:41) ~[adventure-text-serializer-gson-4.17.0.jar:?]
        at Test.jar/com.cavetale.test.TestPlugin.onCommand(TestPlugin.java:71) ~[Test.jar:?]
        at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[paper-api-1.20.6-R0.1-SNAPSHOT.jar:?]
        ... 9 more
```
I hope this helps.

### Steps/models to reproduce

- Put the above code in a test command:
  - Make leather armor item with trim
  - Turn into Adventure hover event
  - Serialize with GsonComponentSerializer
- Issue the command
- Watch console

### Plugin and Datapack List

Bukkit Plugins:
 - AdviceAnimals, AFK, AntiPopup, Area, ArmorStandEditor, Auction, Bans, Blank, BungeeTabListPlus, Chair
Chat, Chess, Connect, Core, Countdown, Crypt, CullMob, Dungeons, Dusk, Easter
Editor, Election, Enemy, Exploits, Fam, FastLeafDecay, Festival, Fly, FreeHat, Halloween
Home, Inventory, ItemStore, KeepInventory, Kit, LinkPortal, MagicMap, Mail, MassStorage, Maypole
MemberList, Menu, Merchant, Money, Mytems, OpenInv, Perm, Photos, PlayerCache, PlayerInfo
PlugInfo, PlugManX, PocketMob, Poster, Protect, RandomPlayerHead, Resident, Resource, ResourcePack, Rules
Server, Shop, Shutdown, Sidebar, SignSpy, Skills, Spawn, Spike, SQL, StarBook
StopRain, Structure, Televator, Test, Ticket, Tinfoil, Title, TooManyEntities, TPA, Trees
Tutor, Vault, VoidGenerator, Vote, Vulcan, Wall, Wardrobe, Warp, Watchman, Webserver
WorldEdit, WorldGuard, WorldMarker, Worlds

There are 3 data pack(s) enabled: [vanilla (built-in)], [file/bukkit (world)], [paper (built-in)]

### Paper version

This server is running Paper version 1.20.6-130-master@5d8e53d (2024-06-01T14:32:53Z) (Implementing API version 1.20.6-R0.1-SNAPSHOT)
You are running the latest version

### Other

- Other leather armor materials cause this as well.
- Other TrimMaterial and TrimPattern combinations cause this as well.
- Remove the line that adds the trim, and everything works just fine.
- This might well be a bug in Adventure rather than Paper; sorry if it is.

--- a/patches/server/0010-Adventure.patch
+++ b/patches/server/0010-Adventure.patch
@@ -1157,7 +1157,7 @@ index 0000000000000000000000000000000000000000..2fd6c3e65354071af71c7d8ebb97b559
 +}
 diff --git a/src/main/java/io/papermc/paper/adventure/PaperAdventure.java b/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
 new file mode 100644
-index 0000000000000000000000000000000000000000..fc6e13e4f2408ccbfa645eae2d7ebf4dcfc21908
+index 0000000000000000000000000000000000000000..7397918cf747bc2352bf5bb112a71e7f6844e0e0
 --- /dev/null
 +++ b/src/main/java/io/papermc/paper/adventure/PaperAdventure.java
 @@ -0,0 +1,478 @@
@@ -1594,7 +1594,7 @@ index 0000000000000000000000000000000000000000..fc6e13e4f2408ccbfa645eae2d7ebf4d
 +
 +        @Override
 +        public @NotNull BinaryTagHolder asBinaryTag() {
-+            return BinaryTagHolder.encode(this.codec.encodeStart(NbtOps.INSTANCE, this.value).getOrThrow(IllegalArgumentException::new), NBT_CODEC);
++            return BinaryTagHolder.encode(this.codec.encodeStart(CraftRegistry.getMinecraftRegistry().createSerializationContext(NbtOps.INSTANCE), this.value).getOrThrow(IllegalArgumentException::new), NBT_CODEC);
 +        }
 +    }
 +
@@ -1622,7 +1622,7 @@ index 0000000000000000000000000000000000000000..fc6e13e4f2408ccbfa645eae2d7ebf4d
 +    // Style
 +
 +    public static net.minecraft.network.chat.Style asVanilla(final Style style) {
-+        final RegistryOps<Object> ops = RegistryOps.create(JavaOps.INSTANCE, CraftRegistry.getMinecraftRegistry());
++        final RegistryOps<Object> ops = CraftRegistry.getMinecraftRegistry().createSerializationContext(JavaOps.INSTANCE);
 +        final Object encoded = AdventureCodecs.STYLE_MAP_CODEC.codec()
 +            .encodeStart(ops, style).getOrThrow(IllegalStateException::new);
 +
@@ -1631,7 +1631,7 @@ index 0000000000000000000000000000000000000000..fc6e13e4f2408ccbfa645eae2d7ebf4d
 +    }
 +
 +    public static Style asAdventure(final net.minecraft.network.chat.Style style) {
-+        final RegistryOps<Object> ops = RegistryOps.create(JavaOps.INSTANCE, CraftRegistry.getMinecraftRegistry());
++        final RegistryOps<Object> ops = CraftRegistry.getMinecraftRegistry().createSerializationContext(JavaOps.INSTANCE);
 +        final Object encoded = net.minecraft.network.chat.Style.Serializer.CODEC
 +            .encodeStart(ops, style).getOrThrow(IllegalStateException::new);
 +
@@ -1641,7 +1641,7 @@ index 0000000000000000000000000000000000000000..fc6e13e4f2408ccbfa645eae2d7ebf4d
 +}
 diff --git a/src/main/java/io/papermc/paper/adventure/WrapperAwareSerializer.java b/src/main/java/io/papermc/paper/adventure/WrapperAwareSerializer.java
 new file mode 100644
-index 0000000000000000000000000000000000000000..e58e8ac574eb60205b17e31589f5a7610c60a4e2
+index 0000000000000000000000000000000000000000..c786ddf0ef19757011452204fd11d24541c39d9e
 --- /dev/null
 +++ b/src/main/java/io/papermc/paper/adventure/WrapperAwareSerializer.java
 @@ -0,0 +1,34 @@
@@ -1661,7 +1661,7 @@ index 0000000000000000000000000000000000000000..e58e8ac574eb60205b17e31589f5a761
 +        if (input instanceof AdventureComponent) {
 +            return ((AdventureComponent) input).adventure;
 +        }
-+        final RegistryOps<Object> ops = RegistryOps.create(JavaOps.INSTANCE, CraftRegistry.getMinecraftRegistry());
++        final RegistryOps<Object> ops = CraftRegistry.getMinecraftRegistry().createSerializationContext(JavaOps.INSTANCE);
 +        final Object obj = ComponentSerialization.CODEC.encodeStart(ops, input)
 +            .getOrThrow(s -> new RuntimeException("Failed to encode Minecraft Component: " + input + "; " + s));
 +        final Pair<Component, Object> converted = AdventureCodecs.COMPONENT_CODEC.decode(ops, obj)
@@ -1671,7 +1671,7 @@ index 0000000000000000000000000000000000000000..e58e8ac574eb60205b17e31589f5a761
 +
 +    @Override
 +    public net.minecraft.network.chat.Component serialize(final Component component) {
-+        final RegistryOps<Object> ops = RegistryOps.create(JavaOps.INSTANCE, CraftRegistry.getMinecraftRegistry());
++        final RegistryOps<Object> ops = CraftRegistry.getMinecraftRegistry().createSerializationContext(JavaOps.INSTANCE);
 +        final Object obj = AdventureCodecs.COMPONENT_CODEC.encodeStart(ops, component)
 +            .getOrThrow(s -> new RuntimeException("Failed to encode adventure Component: " + component + "; " + s));
 +        final Pair<net.minecraft.network.chat.Component, Object> converted = ComponentSerialization.CODEC.decode(ops, obj)
@@ -1829,14 +1829,15 @@ index 0000000000000000000000000000000000000000..8323f135d6bf2e1f12525e05094ffa3f
 +}
 diff --git a/src/main/java/io/papermc/paper/adventure/providers/DataComponentValueConverterProviderImpl.java b/src/main/java/io/papermc/paper/adventure/providers/DataComponentValueConverterProviderImpl.java
 new file mode 100644
-index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0eaf1371ba
+index 0000000000000000000000000000000000000000..6b37c0ebda4a0cfcf9c3b2c3483ffababe622555
 --- /dev/null
 +++ b/src/main/java/io/papermc/paper/adventure/providers/DataComponentValueConverterProviderImpl.java
-@@ -0,0 +1,75 @@
+@@ -0,0 +1,82 @@
 +package io.papermc.paper.adventure.providers;
 +
 +import com.google.gson.JsonElement;
 +import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import com.mojang.serialization.DynamicOps;
 +import com.mojang.serialization.JsonOps;
 +import io.papermc.paper.adventure.PaperAdventure;
 +import java.util.List;
@@ -1849,6 +1850,8 @@ index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0e
 +import net.minecraft.core.registries.BuiltInRegistries;
 +import net.minecraft.nbt.NbtOps;
 +import net.minecraft.nbt.Tag;
++import net.minecraft.resources.RegistryOps;
++import org.bukkit.craftbukkit.CraftRegistry;
 +import org.checkerframework.checker.nullness.qual.NonNull;
 +import org.checkerframework.checker.nullness.qual.Nullable;
 +import org.checkerframework.framework.qual.DefaultQualifier;
@@ -1865,6 +1868,10 @@ index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0e
 +        return ID;
 +    }
 +
++    private static <T> RegistryOps<T> createOps(final DynamicOps<T> delegate) {
++        return CraftRegistry.getMinecraftRegistry().createSerializationContext(delegate);
++    }
++
 +    @SuppressWarnings({"unchecked", "rawtypes"})
 +    @Override
 +    public Iterable<DataComponentValueConverterRegistry.Conversion<?, ?>> conversions() {
@@ -1872,7 +1879,7 @@ index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0e
 +            DataComponentValueConverterRegistry.Conversion.convert(
 +                PaperAdventure.DataComponentValueImpl.class,
 +                GsonDataComponentValue.class,
-+                (key, dataComponentValue) -> gsonDataComponentValue((JsonElement) dataComponentValue.codec().encodeStart(JsonOps.INSTANCE, dataComponentValue.value()).getOrThrow())
++                (key, dataComponentValue) -> gsonDataComponentValue((JsonElement) dataComponentValue.codec().encodeStart(createOps(JsonOps.INSTANCE), dataComponentValue.value()).getOrThrow())
 +            ),
 +            DataComponentValueConverterRegistry.Conversion.convert(
 +                GsonDataComponentValue.class,
@@ -1882,13 +1889,13 @@ index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0e
 +                    if (type == null) {
 +                        throw new IllegalArgumentException("Unknown data component type: " + key);
 +                    }
-+                    return new PaperAdventure.DataComponentValueImpl(type.codecOrThrow(), type.codecOrThrow().parse(JsonOps.INSTANCE, dataComponentValue.element()).getOrThrow(IllegalArgumentException::new));
++                    return new PaperAdventure.DataComponentValueImpl(type.codecOrThrow(), type.codecOrThrow().parse(createOps(JsonOps.INSTANCE), dataComponentValue.element()).getOrThrow(IllegalArgumentException::new));
 +                }
 +            ),
 +            DataComponentValueConverterRegistry.Conversion.convert(
 +                PaperAdventure.DataComponentValueImpl.class,
 +                DataComponentValue.TagSerializable.class,
-+                (key, dataComponentValue) -> BinaryTagHolder.encode((Tag) dataComponentValue.codec().encodeStart(NbtOps.INSTANCE, dataComponentValue.value()).getOrThrow(), PaperAdventure.NBT_CODEC)
++                (key, dataComponentValue) -> BinaryTagHolder.encode((Tag) dataComponentValue.codec().encodeStart(createOps(NbtOps.INSTANCE), dataComponentValue.value()).getOrThrow(), PaperAdventure.NBT_CODEC)
 +            ),
 +            DataComponentValueConverterRegistry.Conversion.convert(
 +                DataComponentValue.TagSerializable.class,
@@ -1899,7 +1906,7 @@ index 0000000000000000000000000000000000000000..ef19891a895e699cac638bd9b9f2fa0e
 +                        throw new IllegalArgumentException("Unknown data component type: " + key);
 +                    }
 +                    try {
-+                        return new PaperAdventure.DataComponentValueImpl(type.codecOrThrow(), type.codecOrThrow().parse(NbtOps.INSTANCE, tagSerializable.asBinaryTag().get(PaperAdventure.NBT_CODEC)).getOrThrow(IllegalArgumentException::new));
++                        return new PaperAdventure.DataComponentValueImpl(type.codecOrThrow(), type.codecOrThrow().parse(createOps(NbtOps.INSTANCE), tagSerializable.asBinaryTag().get(PaperAdventure.NBT_CODEC)).getOrThrow(IllegalArgumentException::new));
 +                    } catch (final CommandSyntaxException e) {
 +                        throw new IllegalArgumentException(e);
 +                    }
@@ -5711,7 +5718,7 @@ index 0000000000000000000000000000000000000000..3aedd0bbc97edacc1ebf71264b310e55
 +}
 diff --git a/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
 new file mode 100644
-index 0000000000000000000000000000000000000000..0b23e6aab901d359845573e12e939a231ee30ada
+index 0000000000000000000000000000000000000000..69b3aa957161e75d6344e437a8b4a4533ef00523
 --- /dev/null
 +++ b/src/test/java/io/papermc/paper/adventure/AdventureCodecsTest.java
 @@ -0,0 +1,403 @@
@@ -5923,7 +5930,7 @@ index 0000000000000000000000000000000000000000..0b23e6aab901d359845573e12e939a23
 +            JavaOps.INSTANCE,
 +            JsonOps.INSTANCE
 +        )
-+            .map(ops -> RegistryOps.create(ops, AbstractTestingBase.REGISTRY_CUSTOM))
++            .map(ops -> AbstractTestingBase.REGISTRY_CUSTOM.createSerializationContext(ops))
 +            .toList();
 +    }
 +

